torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,367 @@
+  }
+}
+
+TensorPipeAgent::~TensorPipeAgent() {
+  shutdown();
+}
+
+void TensorPipeAgent::start() {
+  context_->registerTransport(
+      1, "tcp", std::make_shared<tensorpipe::transport::uv::Context>());
+  context_->registerTransport(
+      0, "shm", std::make_shared<tensorpipe::transport::shm::Context>());
+  context_->registerChannelFactory(
+      1,
+      "basic",
+      std::make_shared<tensorpipe::channel::basic::BasicChannelFactory>());
+  context_->registerChannelFactory(
+      0, "cma", tensorpipe::channel::cma::CmaChannelFactory::create());
+
+  std::vector<std::string> addresses = {createUniqueShmAddr(),
+                                        "tcp://127.0.0.1"};
+
+  listener_ = tensorpipe::Listener::create(context_, addresses);
+
+  // Store our own url.
+  const auto address = listener_->url("tcp");
+  const std::vector<uint8_t> selfAddrData(address.begin(), address.end());
+  addressStore_->set(workerInfo_.name_, selfAddrData);
+
+  for (const auto& p : workerNameToInfo_) {
+    const auto& name = p.first;
+    auto nodeAddrData = addressStore_->get(name);
+    auto nodeAddrStr =
+        std::string((const char*)nodeAddrData.data(), nodeAddrData.size());
+    workerNameToURL_.insert({name, nodeAddrStr});
+  }
+
+  listener_->accept([this](
+                        const tensorpipe::Error& error,
+                        std::shared_ptr<tensorpipe::Pipe> pipe) {
+    onListenerAccepted(error, pipe);
+  });
+
+  isRunning_ = true;
+}
+
+void TensorPipeAgent::onListenerAccepted(
+    const tensorpipe::Error& error,
+    std::shared_ptr<tensorpipe::Pipe>& pipe) {
+  // TODO: Handle listener error
+  if (error) {
+    LOG(WARNING) << "got error from listener: " << error.what();
+    return;
+  }
+
+  // Accept the next connection request
+  listener_->accept([this](
+                        const tensorpipe::Error& error,
+                        std::shared_ptr<tensorpipe::Pipe> pipe) {
+    onListenerAccepted(error, pipe);
+  });
+
+  // Arm for server read
+  respond(pipe);
+}
+
+void TensorPipeAgent::pipeRead(
+    std::shared_ptr<tensorpipe::Pipe> pipe,
+    std::function<void(const tensorpipe::Error&, Message&&)> fn) {
+  pipe->readDescriptor(
+      [this, fn{std::move(fn)}, pipe](
+          const tensorpipe::Error& error, tensorpipe::Message&& tpMessage) {
+        if (error) {
+          fn(error, std::move(Message()));
+          return;
+        }
+
+        // Allocate memory and fill in pointers
+        Message rpcMessage = tensorpipeAllocateMessage(tpMessage);
+        TORCH_INTERNAL_ASSERT(
+            rpcMessage.tensors().size() == tpMessage.tensors.size(),
+            "Tensor num mismatch");
+        tpMessage.data = (uint8_t*)(rpcMessage.payload().data());
+        for (int i = 0; i < rpcMessage.tensors().size(); i++) {
+          auto& rpcTensor = rpcMessage.tensors()[i];
+          auto& tpTensor = tpMessage.tensors[i];
+          tpTensor.data = (uint8_t*)(rpcTensor.data_ptr());
```

Comment:
In `wireSerialization` in `rpc/utils.cpp`, we would send out the entire storage if the the tensor data is a significant portion of the storage. So, there is chance that this tensor might not be contiguous (@jjlilley please correct me if I am wrong). Do we need to check that here?

Reply:
Yes, though in the ``tensorpipeSerialize()`` function rather than here.

The code should probably rpc::cloneSparseTensors() here to avoid that situation... with training data, it's common to send 1 out of 100 rows in an rpc. Locally, it's implemented just as a view into the original data with a stride, but you want to re-copy to avoid 100x blowup in size.

Reply:
I'd like to understand better what the RPC API contract is here. If the agent is allowed to clone sparse tensors and "compress" the storage by only sending over the parts that are "visible", this means that the receiver will get a slightly different tensor than the one that was sent, namely its strides may be different and its `.storage()` may have a different size. Is this intended/expected?

Is there a generic helper function that can be shared by all agents to make sure this conversion is done correctly and consistently? It looks like it becomes complicated really quickly. Think, for example, what would happen if one sends a self-overlapping tensor (e.g., a 2x3 tensor, backed by a storage of size 5, with strides (2, 1), which means that [0,2] is the same location as [1,0]).

Reply:
@lw That's a good point to raise.

- The existing logic was effectively added in December when we realized that we were sending 100x too much data in the wire in the testing training; we were completely saturating the NICs, trainer qps was quite low, and it was quite clear that we had to do address this. It's apparently a common pattern to do this for training samples (have a large tensor of 100 training rows, but sample/send one at a time). 

- Since then, the team has realistically been sufficiently occupied with various training performance issues that rationalizing the api contract honestly hasn't been high pri.

- So, presently, the other rpc_agents do not guarantee that the resulting strides and underlying .storage() are identical. IMO, it would be reasonable to add an default-off option to do so, if there's a need for it. But e.g. the platform in general doesn't even guarantee that .storage() even exists or is usable.

Reply:
This looks like an important thing we need to clarify in our API doc. Let me create an issue to track this.

Reply:
#35766